services/antivirus/pkg/service/service.go

```diff
@@ -0,0 +1,205 @@
+package service
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/cs3org/reva/v2/pkg/bytesize"
+	ctxpkg "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/cs3org/reva/v2/pkg/events"
+	"github.com/cs3org/reva/v2/pkg/events/stream"
+	"github.com/cs3org/reva/v2/pkg/rhttp"
+	"github.com/owncloud/ocis/v2/ocis-pkg/log"
+	"github.com/owncloud/ocis/v2/services/antivirus/pkg/config"
+	"github.com/owncloud/ocis/v2/services/antivirus/pkg/scanners"
+)
+
+// Scanner is an abstraction for the actual virus scan
+type Scanner interface {
+	Scan(file io.Reader) (scanners.ScanResult, error)
+}
+
+// NewAntivirus returns a service implementation for Service.
+func NewAntivirus(c *config.Config, l log.Logger) (Antivirus, error) {
+	av := Antivirus{c: c, l: l, client: rhttp.GetHTTPClient(rhttp.Insecure(true))}
+
+	var err error
+	av.s, err = scanners.New(c.Scanner)
+	if err != nil {
+		return av, err
+	}
+
+	switch o := events.PostprocessingOutcome(c.InfectedFileHandling); o {
+	case events.PPOutcomeContinue, events.PPOutcomeAbort, events.PPOutcomeDelete:
+		av.o = o
+	default:
+		return av, fmt.Errorf("unknown infected file handling '%s'", o)
+	}
+
+	if c.MaxScanSize != "" {
+		b, err := bytesize.Parse(c.MaxScanSize)
+		if err != nil {
+			return av, err
+		}
+
+		av.m = b.Bytes()
+	}
+
+	return av, nil
+}
+
+// Antivirus defines implements the business logic for Service.
+type Antivirus struct {
+	c *config.Config
+	l log.Logger
+	s Scanner
+	o events.PostprocessingOutcome
+	m uint64
+
+	client *http.Client
+}
+
+// Run runs the service
+func (av Antivirus) Run() error {
+	evtsCfg := av.c.Events
+
+	var rootCAPool *x509.CertPool
+	if evtsCfg.TLSRootCACertificate != "" {
+		rootCrtFile, err := os.Open(evtsCfg.TLSRootCACertificate)
+		if err != nil {
+			return err
+		}
+
+		var certBytes bytes.Buffer
+		if _, err := io.Copy(&certBytes, rootCrtFile); err != nil {
+			return err
+		}
+
+		rootCAPool = x509.NewCertPool()
+		rootCAPool.AppendCertsFromPEM(certBytes.Bytes())
+		evtsCfg.TLSInsecure = false
+	}
+
+	stream, err := stream.NatsFromConfig(stream.NatsConfig(av.c.Events))
+	if err != nil {
+		return err
+	}
+
+	ch, err := events.Consume(stream, "antivirus", events.StartPostprocessingStep{})
+	if err != nil {
+		return err
+	}
+
+	for e := range ch {
+		ev := e.Event.(events.StartPostprocessingStep)
+		if ev.StepToStart != events.PPStepAntivirus {
+			continue
+		}
+
+		var errmsg string
+		res, err := av.process(ev)
+		if err != nil {
+			errmsg = err.Error()
+		}
+
+		outcome := events.PPOutcomeContinue
+		if res.Infected {
+			outcome = av.o
+		}
```

Comment:
on error the res is still used to set the outcome, which has res.Infected set to false ...

Reply:
refresh the browser :smile: